api/agent/agent.go

```diff
@@ -108,14 +109,15 @@ type Agent interface {
 	// TODO this should be derived from our metrics
 	Stats() Stats
 
-	// Return the http.Handler used to handle Prometheus metric requests
+	AddCallListener(extenders.CallListener)
```

Comment:
I think it would be nice if we made something that implements `agent.Agent` interface to do this. obviously, it's thin, but this is just one-more-thing and there will be other one-more-things to tack onto the `*agent.agent` (lower case... sigh), so it would be nice to avoid one-more-thing where easily possible. e.g.

```go
type listenerAgent struct {
  agent.Agent
  cls []extenders.CallListener
}

// don't implement GetCall, use underlying Agent's

func (l *listenerAgent) Submit(call Call) error {
  m := call.Model()
  for _, cl := range l.cls {
    cl.BeforeCall(ctx, m)
  }

  // actual agent submit
  err := l.Agent.Submit(call)

  m := call.Model()
  for _, cl := range l.cls {
    cl.AfterCall(ctx, m)
  }
}

// agent/agent.go initialization
func New(...) agent.Agent {
  agent = newAgent(...)
  return listener.New(agent)
}
```

this also means that users aren't expressly calling methods against the `Agent` (don't like exposing this, want to hide it), the `*server.Server` can expose `AddCallListener`. one downside i guess is maneuvering the timeout in the `context.Context` here, would need to move it from where we're currently setting it (to account for `BeforeCall`, specifically)

anyway, possibly a terrible idea, also willing to punt, just throwing ideas around.

Reply:
i guess the other thing this doesn't account for is that BeforeCall will be called before finding a 'slot', so BeforeCall has a higher chance of being called even when the call doesn't run (maybe not an issue? idk)

Reply:
leaving, but kind of leaning on just shoving into `call.Start` and `call.End`, that was the idea of those methods anyway. it brings up a point that we can do very similar to what's proposed here with agent with the `call.Call` interface, i.e. implement the `call.Call` interface (`Start`, `End`) with the `extender.CallListener` stuff, which makes sense too (maybe plumbing challenge?). just throwing ideas around again.

Reply:
I tried to leave this out of the Agent interface, but couldn't see a way to do it. And I'm guessing there'll be more callbacks within the agent down the road so I think we'll just have to live with it having these things. Which also makes breaking them apart, maybe kind of pointless?

Regarding shoving into call.Start() and call.End(), do you mean pass the listeners into those functions or pass the agent in?  Maybe we can figure out the optimal spot as a follow up. 

And it looks like it finds the slot right before the listeners are called?

Reply:
github reviews collapsing + beehive email delays will be the death of me (coronary). it probably makes sense to put the agent on the call for it's fields, we're atm selectively dropping in the datastore and some friends. makes sense just passing in additional fields we need so that nobody tries anything too weird, as well. i don't like my `listenerAgent` idea I left here ex-post-facto, using `Start` and `End` as we are now seems much better, but yea, think we should shove `[]CallListeners` on the `agent.Call`